### ISSUE TITLE
web application is not working: websocket? port 80?

### ISSUE THREAD

**Issue by danielecr** [zulip/docker-zulip#215]
I experienced this problem from the begin: message sent from web application are not working.

I changed the -compose.yml file to be used in a docker in swarm mode with traefik routing traffic the service, also port 80 is not exposed at all outside.

I see now that web application rely on websocket, I ispected the messages exchanded by looking at chromium devtools. The sent message looks correct. I looked at received message I see just 'o'.

Does it works? I sent a message 1 week ago, I never see it till now. There is something wrong in backend?

Not so often I receive email from Django about error log. I paste one of those here:

~~~
Logger django.pylibmc, from module django_pylibmc.memcached line 132:
Error generated by Anonymous user (not logged in) on a1d3a3aa462d deployment

Traceback (most recent call last):
  File "/home/zulip/deployments/2019-06-20-00-28-26/zulip-py3-venv/lib/python3.5/site-packages/django/contrib/sessions/backends/base.py", line 202, in _get_session
    return self._session_cache
AttributeError: 'SessionStore' object has no attribute '_session_cache'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/zulip/deployments/2019-06-20-00-28-26/zulip-py3-venv/lib/python3.5/site-packages/django_pylibmc/memcached.py", line 130, in get
    return super(PyLibMCCache, self).get(key, default, version)
  File "/home/zulip/deployments/2019-06-20-00-28-26/zulip-py3-venv/lib/python3.5/site-packages/django/core/cache/backends/memcached.py", line 79, in get
    val = self._cache.get(key)
pylibmc.ConnectionError: error 3 from memcached_get(:1:django.contrib.sessions.cache): (0x3a378c0) CONNECTION FAILURE(Connection reset by peer),  host: memcached:11211 -> libmemcached/io.cc:466


Deployed code:
- ZULIP_VERSION: 2.0.4
- version: docker


Request info: none
~~~

**Comment by timabbott** [comment]
From that exception, it sounds like Docker Swarm restarted (or shut down) the memcached server that Zulip was connecting to and that broke the connections from Zulip to memcached; restarting the Zulip service should fix that.  If you haven't done that, it's potentially the cause of the problem -- Zulip requires memcached to function.

Can you provide more detail on what you mean by message sending not working?  This project certainly works in general.  Do most messages send but you had one that way delayed, or does nothing send?  Etc.

**Comment by danielecr** [comment]
First of all, question. I mean the web application does not work. And that is true also for the electron based application (desktop). But it works from the android app, and probably iOS too (I have to ask details to my collegue who is testing it, but he reported no problem).

I checked and the services are up and running since 12 days ago. I checked the log and all are the same container (not only service, but container is running since 12 days). While the message is dated september 8th, yesterday, at 8:02. I logged into zulip container (with docker exec) and tried `ping memcached`, it works. `ping zulip_memcached`, it works. Both solved with the same ip. I checked for possible conflict, nothing found.

But I looked at log of zulip at that time and I see a lot of error, and process respawn, on the same zulip container.

I am not a Django expert, there is a simple python code to test if memcached connection is working, other than ping to the service?

I can send docker logs, but there is other logs elsewhere? from zulip I mean, what I sent initially comes from an email notification

**Comment by danielecr** [comment]
I think relevant logs are this:
~~~
2019-09-09 10:28:25,394 INFO exited: process-fts-updates (exit status 1; not expected)
2019-09-09 10:28:26,427 INFO spawned: 'process-fts-updates' with pid 263
2019-09-09 10:28:28,066 INFO success: process-fts-updates entered RUNNING state, process has stayed up for > than 1 seconds (startsecs)
2019-09-09 12:28:34,350 INFO exited: process-fts-updates (exit status 1; not expected)
2019-09-09 12:28:35,357 INFO spawned: 'process-fts-updates' with pid 294
2019-09-09 12:28:36,495 INFO success: process-fts-updates entered RUNNING state, process has stayed up for > than 1 seconds (startsecs)
~~~

I can not say if this project works elsewhere, with other configuration, but for sure it does not send messages from the web application, and also not from the electron based linux application, with my current configuration. Most of the messages are droped, "some" but very few are delivered, I do not know how much/when/why.

I do not think to have used a wrong configuration, it is the same as the composer provided in this repository, but used in swarm mode, so I did pulled the image, and not built.

**Comment by danielecr** [comment]
it may be this
https://stackoverflow.com/questions/45464191/traefik-configuration-to-allow-websocket
traefik bug with websocket before 1.3.8 (I am running 1.3.7)

**Comment by danielecr** [comment]
this label is needed for traefik:
```
 - "traefik.wss.protocol=https"
```
it is for websocket, obvious to zulip dev people ... can this be added to the README? or elsewhere?

**Comment by timabbott** [comment]
Where would you look for it in the docs?  

If you want to contribute traefik config for https://zulip.readthedocs.io/en/latest/production/deployment.html#putting-the-zulip-application-behind-a-reverse-proxy, that might be best.

**Comment by danielecr** [comment]
well, it was a very tricky error in traefik.backend label setup:
I wrote zulip there, but from the point of view of traefik, zulip is the service stack name, the target service is really zulip_zulip. So
> `- "traefik.backend=zulip_zulip"`

fixed the problem. I do not know exactly the policy for addressing zulip in docker swarm, but maybe there is a rotation on services composing the stack (it is a nonsense, but it may explain something about my logs)
Now I see that in https://github.com/containous/traefik/issues/4704 the `traefik.backend` is omitted and in this way the service name (where labels are defined) is taken as target backend.

To sum it up:
1. use `insecureSkipVerify = true` in traefik configuration file
2. then use https protocol (port 443)
3. be aware to specify the right backend name (and not the stack name) or omit completely from the traefik labels

**Comment by timabbott** [comment]
Thanks for the follow-up!  It'd be great to document what you needed to do here to help future users.  

@danielecr would you be up for submitting your traefik configuration as a sample here: https://zulip.readthedocs.io/en/latest/production/deployment.html#putting-the-zulip-application-behind-a-reverse-proxy ?  

The source is in https://github.com/zulip/zulip/ in `docs/production/deployment.md`.
